feat(uac): allow user select default ownership for resources

olljanat opened the pull request:
Allow user to switch back to old behavior which was there before breaking changes on #2137 

Fixes #685

olljanat commented:
@deviantony when you have time it would be nice to get some feedback to this one?

deviantony commented:
Yes I'll try to review ASAP

olljanat commented:
> I've enabled it, created a container on the CLI and it was still declared as administrator restricted in Portainer.

Actually it looks to be functional but if you login as admin it is just shown incorrectly on UI. You need login as normal user to be able see difference.

Anyway, included DB migration step now to simplify testing and corrected that tooltip text. I will figure out how to make UI shown ownership for admins correctly later.

olljanat commented:
@deviantony now ownership is shown correctly for admin users too.

olljanat commented:
@chiptus @Louis-Philippe @kendrick can someone of you test this one and give me feedback that what else I probably have forgot?

It make me very nervous that we are not able to update to any version after 1.19.1 before this one is merged and I really would like to get those new features which are coming to 1.20 to use.

deviantony commented:
I was able to test this again.

There is something confusing here, we enable a switch titled "Make externally created resources public by default" but it will also switch to public any existing administrator restricted resource.

I don't think that this is the behavior that we want integrate. cc @ncresswell for another review (you can use `portainer/portainer:pr2424`).

olljanat commented:
@deviantony OMG. I can see it also with that image. Very weird as based on code ( https://github.com/olljanat/portainer/blob/0744b0517ad09da80f3c95451e25efe25358a58d/api/http/proxy/access_control.go#L47 ) resources should be public only if *Make externally created resources public by default* is enable and resourceControl is nil

I will troubleshoot it more tomorrow

olljanat commented:
@ncresswell yes target is to let user choose between current behavior an the old one which still exists on 1.19.1 version.

@deviantony I found that DB migration introduced by #2137 actually **removes** resourceControl from all services which are set to administrator only. 
https://github.com/portainer/portainer/blob/e1e263d8c88f0eefa68f4e013a0ce5d6b24f56c3/api/bolt/migrator/migrate_dbversion13.go#L11

So I need create DB migration which re-creates them and on it must be also created for all new resources. Only then switching between *Make externally created resources public by default* setting modes can work correctly. 

Is there some way detect on code that which resources are created outside of Portainer?

Ping @ricardona any comments?

ricardona commented:
> Is there some way detect on code that which resources are created outside of Portainer?

@deviantony in #960:

>The database only store UAC related information, it does not store any Docker related information (containers, etc...). So that would require parsing all the containers running on boot / when adding an endpoint and automatically add information on each resource. And also a refactor on how the resource control model is currently defined.

olljanat commented:
@deviantony I modified code now on the way that all resources which are created from Portainer will have ResourceControl no matter who is owner. That way *Make externally created resources public by default* setting will only affect resources which are created outside of Portainer.

I also disabled deleting administrators resource controls which was introduced on DBVersion 14 so anyone who are upgrading from version < 1.19.2 will avoid that. **But** because some users are already on 1.19.2 there is now also popup telling to user that some resources might be changed to public and user should review them.

Way how external resources are now handled is that API creates resource control with type *DynamicResourceControl* for them on fly and that is these are not stored to database. Idea is that it can be also used with #1257 when it is implemented.

Known issue is that UI show ownership incorrectly on Stacks page but it would be nice to hear your feedback if this is now over all on right track before I dig into these minor things?

cc @ricardona as I needed to change logic which you implemented.

olljanat commented:
> The Make externally created resources public by default switch will actually trigger an update of the settings right after clicking on it whereas for all the other switches in this view we expect the user to click on the "Save settings" button. In order to keep a consistent UX (albeit a bit worst) I think we should simply expect the user to click on save settings for this switch as well.

Fixed.

> Please add these changes to the UX and we'll try to go forward to bring this in 1.20.

@deviantony Do you have idea what would be best way to solve ownership which is shown on https://github.com/portainer/portainer/blob/develop/app/portainer/components/datatables/stacks-datatable/stacksDatatable.html#L74 ?

Stacks looks to be more tricky than everything else as there is multiple ways how stacks are resolved to list. (rights works already correctly but ownership is shown incorrectly).

olljanat commented:
@deviantony FYI, rebased to latest state of develop and fixed merge conflicts.

olljanat commented:
@deviantony new year and same old issues. Any possibility to look this one on near weeks? We really would like to rid of known bugs which exists on 1.19.1 and are fixed on later version but cannot do upgrade before this on is merged.

deviantony commented:
@olljanat Happy new year :-)

We'll try our best to make this part of the next major release.

olljanat commented:
@deviantony is there any schedule for that next major release? (just that I know to do rebase for this when you are ready to continue with it)

olljanat commented:
@deviantony FYI. I did rebase for old commit and added one more which tries to fix those minor issues as good as I was able to.

How ever I handled external stack ownership displaying issue on way that I removed it totally. Would that be good enough? (I really would like to get this one merged). Here is picture how it looks like:
![image](https://user-images.githubusercontent.com/6213926/56097492-adf8a080-5efd-11e9-9d40-caed3bddb896.png)

deviantony commented:
@olljanat thanks for the rebase, this is still in our considerations and we really want this built-in too. I'm really sorry for the lack of update on this topic but our resources are limited as you know. Ping me on slack so that we can discuss this topic a bit more if you want.

olljanat commented:
Closing as new RBAC extension (which will be released with next Portainer version) will fix #685
